src/txn/txn_timestamp.c

```diff
@@ -187,6 +187,8 @@ __txn_get_pinned_timestamp(
 	/* Look for the oldest ordinary reader. */
 	__wt_readlock(session, &txn_global->read_timestamp_rwlock);
 	TAILQ_FOREACH(txn, &txn_global->read_timestamph, read_timestampq) {
+		/* Copy value out to prevent possible race condition. */
+		tmp_read_timestamp = txn->read_timestamp;
```

Comment:
The compiler can optimize this assignment away (similarly below). You could either add an explicit read barrier or mark the variable declaration volatile (https://barrgroup.com/Embedded-Systems/How-To/C-Volatile-Keyword).

This is subtle - I'd be happy to talk through the reasons and alternatives tomorrow if you'd like.

Reply:
I've pushed up a change which uses read barriers, let me know what you think. I'd be interested in having a discussion about them though tomorrow.